pkg/framework/plugins/defaultevictor/defaultevictor.go

```diff
@@ -71,6 +72,32 @@ func New(args runtime.Object, handle framework.Handle) (framework.Plugin, error)
 	ev.handle = handle
 	ev.args = defaultEvictorArgs
 
+	ev.constraints = append(ev.constraints, func(pod *v1.Pod) error {
```

Comment:
The default evictor is not responsible for filtering out pods. Only the performing checks whether a pod can be evicted. Printing any information about excluding/including pods from processing by a strategy needs to be printed by the filtering code.

Reply:
Also, assuming the case of a cluster with thousands of pods getting excluded, there would be thousands of log lines produced by each strategy saying "pod 'POD' excluded from eviction by 'filter' extension point".

Reply:
@migueleliasweb each strategy can have its own custom filters which may be excluding excessive number of pods. Even with namespace exclusion getting logged, there may be other unreported places of pods getting filter out.

Reply:
Filters are meant to be invoked multiple times. Whereas eviction only once. That's why it is practical to log reasons why a pod can not be evicted. Whereas filters are intended to reduce the number of pods getting processed by each strategy.

Reply:
Also, if a pod that's excluded/included based on a namespace gets here, something's wrong. The only exception arethe utilization strategies which currently do not perform the namespace inclusion/exclusion check.

Reply:
Thanks for the feedback, @ingvagabund .

> Also, assuming the case of a cluster with thousands of pods getting excluded, there would be thousands of log lines produced by each strategy saying "pod 'POD' excluded from eviction by 'filter' extension point".

> Also, if a pod that's excluded/included based on a namespace gets here, something's wrong. The only exception are the utilization strategies which currently do not perform the namespace inclusion/exclusion check.

I reckon that might be already happening as for each 'cycle check' from the descheduler. At least that's what I'm seeing when I run it on my cluster that has around 1k pods running. As you can see, from the initial issue I oppened (https://github.com/kubernetes-sigs/descheduler/issues/970), there is a namespace constraint (redacted to `foo`) but pods from `kube-system` are shown in the logs. That means the `defaultevictor` is picking up **all pods** and outputting information about whether they shuold be evicted or not.

The core problem here is that there must be a balance between providing useful information regarding **all the pods** in the cluster and why they are or aren't being considered for eviction and the amount of log that is generated by the `Descheduler`. That's what I'm trying to solve here.

A orthogonal idea would be to provide a way to query the `Descheduler` about a specific POD and wether it matches any of the configured policies, instead of outputting tons of logs about all pods in every cycle.

Reply:
> As you can see, from the initial issue I oppened (https://github.com/kubernetes-sigs/descheduler/issues/970), there is a namespace constraint (redacted to foo) but pods from kube-system are shown in the logs. That means the defaultevictor is picking up all pods and outputting information about whether they shuold be evicted or not.

I am not able to reproduce this scenario with the following policy:
```
apiVersion: "descheduler/v1alpha1"
kind: "DeschedulerPolicy"
strategies:
  "PodLifeTime":
    enabled: true
    params:
      podLifeTime:
        maxPodLifeTimeSeconds: 10
      labelSelector:
        matchLabels:
          role: web
      namespaces:
        include:
        - foo
```

and following RS created in kube-system and foo namespaces:
```
apiVersion: apps/v1
kind: ReplicaSet
metadata:
  name: web
  labels:
    env: dev
    role: web
spec:
  replicas: 5
  selector:
    matchLabels:
      role: web
  template:
    metadata:
      labels:
        role: web
    spec:
      nodeSelectorTerms:
      topologySpreadConstraints:
      - maxSkew: 1
        topologyKey: "topology.kubernetes.io/zone"
        whenUnsatisfiable: DoNotSchedule
        labelSelector:
          matchLabels:
            role: web
      containers:
      - name: nginx
        image: gcr.io/google-samples/hello-app:2.0
```

Always getting only pods from foo NS evicted. Pods in the kube-system NS are untouched.

@migueleliasweb would you please provide exact steps on how to reproduce the situation with full descheduler logs?

Reply:
> Since I saw the defaultevictor was being used to provide debugging logs about the running pods, I thought I could just add the log to it.

The changes you performed in this PR are located in the `Filter` extension point which can be disabled if a user decides to provide a custom implementation of the extension point.

I am not opposed to adding more logs (with higher log level) printing all excluded pods. Nevertheless, the code changes needs to go into [BuildGetPodsAssignedToNodeFunc](https://github.com/kubernetes-sigs/descheduler/blob/c10e6d6c129b0db29415fb2043e4b8b07532ecab/pkg/descheduler/pod/pods.go#L151) where the individual pods are included. You will not be able to tell whether a pod was excluded due to namespace exclusion/inclusion. Though, you will be able to see a pod was filter out before it was processed by a strategy.

Reply:
> Always getting only pods from foo NS evicted. Pods in the kube-system NS are untouched.

I think this might have been a misunderstanding. I was just saying the `descheduler` outputs logs from **all pods** but not necesarrily touching/evicting any of them. That's why I added the log lines to the same place (`defaultevictor`).

I will have a look at [BuildGetPodsAssignedToNodeFunc](https://github.com/kubernetes-sigs/descheduler/blob/c10e6d6c129b0db29415fb2043e4b8b07532ecab/pkg/descheduler/pod/pods.go#L151) to see how this could be implemented. It could be interesting to add an even higher level of verbosity for this.

Reply:
Hummm, that's interesting. Yeah you're right. It can be called multiple times. I just noticed there are specific calls for each "plugin". I reckon if we add the plugin information to the option, we could then then output smarter log lines for specific plugins.

From https://github.com/kubernetes-sigs/descheduler/blob/master/pkg/framework/plugins/podlifetime/pod_lifetime.go#L59
```
podFilter, err := podutil.NewOptions().
	WithFilter(podutil.WrapFilterFuncs(handle.Evictor().Filter, handle.Evictor().PreEvictionFilter)).
	WithNamespaces(includedNamespaces).
	WithoutNamespaces(excludedNamespaces).
	WithLabelSelector(podLifeTimeArgs.LabelSelector).
	WithPlugin(PluginName, plugin) # <-- this
	BuildFilterFunc()
if err != nil {
	return nil, fmt.Errorf("error initializing pod filter function: %v", err)
}
```

So the log line could contain something like: 

```
pod <foo> is excluded from namespace <bar> for plugin <taz>
```

With that, we wouldn't have to implement caches or any other fancy ways to get around the fact the `FilterFunc` is called multiple times per cycle.

Reply:
Even a single plugin can have multiple invocations of the filter.

Reply:
Wouldn't that be fine? I guess we still want to output per pod.

Reply:
When the filter is run over the same pod multiple times, the filter will produce duplicated log lines. Also, the namespace filtering is as [third in the list](https://github.com/kubernetes-sigs/descheduler/blob/master/pkg/descheduler/pod/pods.go#L101-L115). So when either `o.labelSelector` or `o.filter` (which is a combination of other filters) gives false, there's not going to be any log saying a pod was excluded from eviction because of the namespace filtering constraint.

If the overall goal is to improve visibility of the decision making to a point of getting complete information about why a pod was excluded from eviction, the `BuildFilterFunc` is not the suitable place for it.